app/controllers/products_controller.rb

```diff
@@ -0,0 +1,68 @@
+class ProductsController < ApplicationController
+  before_action :set_product, only: [:show, :edit, :update, :destroy]
+  before_action :authenticate_admin!, except: [:index, :show]
+
+  # GET /products
+  def index
+    @products = Product.all
+  end
+
+  # GET /products/1
+  def show
+  end
+
+  # GET /products/new
+  def new
+    @product = Product.new
+  end
+
+  # GET /products/1/edit
+  def edit
+  end
+
+  # POST /products
+  def create
+    @product = Product.new(product_params)
+
+    if @product.save
+      redirect_to @product, notice: 'Product was successfully created.'
+    else
+      render :new
+    end
+  end
+
+  # PATCH/PUT /products/1
+  def update
+    if @product.update(product_params)
+      redirect_to @product, notice: 'Product was successfully updated.'
+    else
+      render :edit
+    end
+  end
+
+  # DELETE /products/1
+  def destroy
+    @product.destroy
```

Comment:
Have the else part here as well, what if destroying fails.
At least notify the user.

Reply:
Changes are done.